xray/commands/curation/analyse.go

```diff
@@ -0,0 +1,170 @@
+	}
+	for _, node := range graph.Nodes {
+		packageUrl, name, version := getUrlNameAndVersionByTech(nc.tech, node.Id, nc.url, nc.repo)
+		resp, _, err := nc.artiManager.Client().SendHead(packageUrl, &nc.httpClientDetails)
+		if err != nil || resp.StatusCode >= 400 {
+			if resp == nil {
+				return err
+			}
+			if resp.StatusCode != http.StatusForbidden {
+				log.Error(fmt.Sprintf("Failed to head package %s for download url: %s, status-code: %v, err: %v", node.Id, packageUrl, resp.StatusCode, err))
+				continue
+			}
+		}
+		if resp.StatusCode == http.StatusForbidden {
+			err = nc.addBlockedPackage(packageUrl, parent, parentVersion, respStatus, name, version)
+			if err != nil {
+				log.Error(fmt.Sprintf("%v", err))
+				continue
+			}
+		}
+
+		if err := nc.recursiveNodeCuration(node, respStatus, parent, parentVersion, false); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (nc *treeAnalyzer) addBlockedPackage(packageUrl string, parent, parentVersion string, respStatus *[]PackageStatus, name string, version string) error {
+	getResp, respBody, _, err := nc.artiManager.Client().SendGet(packageUrl, true, &nc.httpClientDetails)
+	if err != nil {
+		if getResp == nil {
+			return err
+		}
+		if getResp.StatusCode != http.StatusForbidden {
+			log.Error(fmt.Sprintf("Failed to get package %s, version: %s for download url: %s. err: %v", name, version, packageUrl, err))
+			return nil
+		}
+	}
+	if getResp.StatusCode == http.StatusForbidden {
+		respError := &utils2.ErrorsResp{}
+		if err := json.Unmarshal(respBody, respError); err != nil {
+			log.Error(fmt.Sprintf("failed to decode response for 'forbidden' response, err: %v", err))
+			return nil
+		}
+		if len(respError.Errors) == 0 {
+			log.Error(fmt.Sprintf("No errors messages for package %s, version: %s for download url: %s ", name, version, packageUrl))
+			return nil
+		}
+		if strings.Contains(strings.ToLower(respError.Errors[0].Message), BlockMessageKey) {
+			depRelation := "indirect"
+			if parent == "" {
+				depRelation = "direct"
+				parent = name
+				parentVersion = version
+			}
+			blockingReason := BlockingReasonPolicy
+			if strings.Contains(strings.ToLower(respError.Errors[0].Message), NotBeingFoundKey) {
+				blockingReason = BlockingReasonNotFound
+			}
+			policies := extractPoliciesFromMsg(respError)
+			*respStatus = append(*respStatus, PackageStatus{
+				PackageName:       name,
+				PackageVersion:    version,
+				BlockedPackageUrl: packageUrl,
+				Action:            Blocked,
+				Policy:            policies,
+				ParentName:        parent,
+				ParentVersion:     parentVersion,
+				BlockingReason:    blockingReason,
+				DepRelation:       depRelation,
+				PkgType:           string(nc.tech),
+			})
+		}
+	}
+	return nil
+}
+
+// / message structure: Package %s:%s download was blocked by JFrog Packages Curation service due to the following policies violated {%s, %s},{%s, %s}.
```

Comment:
Our comments guidelines are the following:
1. Must start with // followed by a white-space 
2. Must start with an upper-case letter.
3. Must have correct English grammar
4. If needed, explain every argument and return vakue.